### PR TITLE
[base] Levenshtein DFA.

### PR DESCRIPTION
**Pull request by ygorshenin**
@mpimenov PTAL

### CODE REVIEW

File: base/levenshtein_dfa.hpp [hunk 1]
```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include "base/string_utils.hpp"
+
+#include "std/cstdint.hpp"
+#include "std/unordered_set.hpp"
+
+namespace strings
+{
+// This class represends a DFA recognizing a language consisting of
```
Comment (ygorshenin) [inline base/levenshtein_dfa.hpp]: Done

File: base/levenshtein_dfa.hpp [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+// all words that are close enough to a given string, in terms of
+// Levenshtein distance. Levenshtein distance treats deletions,
+// insertions and replacements as errors, transpositions are not
+// handled now. The code is based on the work "Fast String Correction
+// with Levenshtein-Automata" by Klaus U. Schulz and Stoyan Mihov.
+// For a fixed number of allowed errors and fixed alphabet the
+// construction time and size of automata is O(length of the pattern),
+// but the size grows exponentially when number of errors grows, so be
```
Comment (ygorshenin) [inline base/levenshtein_dfa.hpp]: Done

File: base/levenshtein_dfa.hpp [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+// reasonable and don't use this class when number of errors is too
```
Comment (ygorshenin) [inline base/levenshtein_dfa.hpp]: Done

File: base/levenshtein_dfa.hpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+public:
+  static size_t const kStartingState;
+  static size_t const kRejectingState;
+  
+  struct Position
+  {
+    Position() = default;
+    Position(size_t offset, uint8_t numErrors);
+
+    bool SumsumedBy(Position const & rhs) const;
```
Comment (ygorshenin) [inline base/levenshtein_dfa.hpp]: Done

File: base/levenshtein_dfa.cpp [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+  }
+  m_alphabet.push_back(missed);
+
+  queue<State> states;
+  map<State, size_t> visited;
+
+  auto pushState = [&states, &visited, this](State const & state, size_t id)
+  {
+    ASSERT_EQUAL(id, m_transitions.size(), ());
+    ASSERT(visited.count(state) == 0, (state, id));
```
Comment (ygorshenin) [inline base/levenshtein_dfa.cpp]: Done

File: base/levenshtein_dfa.cpp [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+  m_alphabet.assign(s.begin(), s.end());
+  my::SortUnique(m_alphabet);
+
+  UniChar missed = 0;
+  for (size_t i = 0; i < m_alphabet.size() && missed >= m_alphabet[i]; ++i)
+  {
+    if (missed == m_alphabet[i])
+      ++missed;
+  }
+  m_alphabet.push_back(missed);
```
Comment (ygorshenin) [inline base/levenshtein_dfa.cpp]: Да.

**Comment by ygorshenin** [comment]
PTAL

File: base/levenshtein_dfa.cpp [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+
+    if (p.m_numErrors == m_maxErrors)
+      return;
+
+    if (p.m_offset + 1 < m_size)
+    {
+      size_t i;
+      if (FindRelevant(p, c, i))
+      {
+        ASSERT(i != 0, (i));
```
Comment (mpimenov) [inline base/levenshtein_dfa.cpp]: ASSERT_NOT_EQUAL

Comment (ygorshenin) [inline base/levenshtein_dfa.cpp]: ASSERT_GREATER. Done.

**Comment by ygorshenin** [comment]
PTAL

**Comment by mpimenov** [comment]
LGTM

**Pull request merged**